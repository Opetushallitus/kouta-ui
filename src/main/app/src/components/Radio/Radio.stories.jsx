import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import Radio, { RadioGroup } from './index';

storiesOf('Radio', module)
  .add('Basic', () => (
    <>
    <Radio value="radioA" onChange={action('change')}>Radio A</Radio>
    <Radio value="radioB" onChange={action('change')}>Radio B</Radio>
    </>
  ))
  .add('With disabled', () => <Radio checked={true} disabled>Radio</Radio>)
  .add('With radio group', () => (
    <RadioGroup value="b" onChange={action('change')}>
      <Radio value="a">Radio A</Radio>
      <Radio value="b">Radio B</Radio>
      <Radio value="c">Radio C</Radio>
    </RadioGroup>
  ))
  .add('With radio group options prop', () => (
    <RadioGroup
      options={[
        { value: 'a', label: 'Radio A' },
        { value: 'b', label: 'Radio B' },
        { value: 'c', label: 'Radio C' },
      ]}
      onChange={action('change')}
    />
  ));
