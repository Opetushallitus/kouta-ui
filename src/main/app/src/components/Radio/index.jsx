import React from 'react';
import styled, { css } from 'styled-components';

import { getThemeProp } from '../../theme';
import { isArray } from '../../utils';

const Label = styled.label`
  font-family: ${getThemeProp('typography.fontFamily')};
  font-size: 1rem;
  display: flex;
  line-height: 1.5;
  color: ${getThemeProp('palette.text.primary')};
`;

const RadioContainer = styled.div`
  ${({ last }) =>
    !last &&
    css`
      margin-bottom: ${({ theme }) => theme.spacing.unit * 0.5}px;
    `}
`;

const LabelWrapper = styled.div`
  flex: 1;
  margin-left: 6px;
`;

const RadioWrapper = styled.div`
  flex: 0;
`;

const RadioInput = styled.input.attrs({ type: 'radio' })`
  font-size: inherit;
`;

export const Radio = ({ children = null, ...props }) => (
  <Label>
    <RadioWrapper>
      <RadioInput type="radio" {...props} />
    </RadioWrapper>
    <LabelWrapper>{children}</LabelWrapper>
  </Label>
);

export const RadioGroup = ({
  value,
  onChange,
  disabled = false,
  options,
  ...props
}) => {
  let children = null;

  if (props.children) {
    const childrenCount = React.Children.count(props.children);

    children = React.Children.map(props.children, (child, index) => {
      const checked = value !== undefined && child.props.value === value;
      const element = React.cloneElement(child, {
        checked,
        onChange,
        disabled,
      });
      const last = index === childrenCount - 1;

      return <RadioContainer last={last}>{element}</RadioContainer>;
    });
  } else if (isArray(options)) {
    children = options.map(({ value: optionValue, label }, index) => (
      <RadioContainer last={index === options.length - 1}>
        <Radio checked={value !== undefined && value === optionValue} onChange={onChange} value={optionValue}>
          {label}
        </Radio>
      </RadioContainer>
    ));
  }

  return <div {...props}>{children}</div>;
};

export default Radio;
