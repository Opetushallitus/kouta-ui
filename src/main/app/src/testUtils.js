import React from 'react';
import { shallow, mount } from 'enzyme';
import { ThemeProvider } from 'styled-components';

import defaultTheme from './theme';

const makeFnWithTheme = fn => tree => {
  return fn(<ThemeProvider theme={defaultTheme}>{tree}</ThemeProvider>);
};

export const shallowWithTheme = makeFnWithTheme(shallow);

export const mountWithTheme = makeFnWithTheme(mount);

export const setMockDate = date => {
  const _Date = Date;

  global.Date.now = () => +date;

  return () => {
    global.Date = _Date;
  };
};
