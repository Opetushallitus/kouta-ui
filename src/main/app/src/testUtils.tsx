import React, { PropsWithChildren } from 'react';

import { configureStore } from '@reduxjs/toolkit';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { ThemeProvider } from 'styled-components';

import { createRootReducer } from './state/rootReducer';
import defaultTheme from './theme';

const makeFnWithTheme = fn => tree => {
  return fn(<ThemeProvider theme={defaultTheme}>{tree}</ThemeProvider>);
};

export const renderWithTheme = makeFnWithTheme(render);

export const mountWithTheme = tree => {
  const { container } = makeFnWithTheme(render)(tree);
  return container.firstChild;
};

export const setMockDate = date => {
  const _Date = Date;

  global.Date.now = () => Number(date);

  return () => {
    global.Date = _Date;
  };
};

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = configureStore({ reducer: createRootReducer({}), preloadedState }),
    ...renderOptions
  }
) {
  // eslint-disable-next-line @typescript-eslint/ban-types
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return (
      <Provider store={store}>
        <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
      </Provider>
    );
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export * from '@testing-library/react';
