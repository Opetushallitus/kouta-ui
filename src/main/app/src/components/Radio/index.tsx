import * as React from 'react';
import styled, { css } from 'styled-components';
import { isArray } from 'lodash';

import Radio from '@opetushallitus/virkailija-ui-components/Radio';

export { default } from '@opetushallitus/virkailija-ui-components/Radio';

export { default as Radio } from '@opetushallitus/virkailija-ui-components/Radio';

// TODO: Move this fixed RadioGroup component to virkailija-ui-components
// This code fixes a bug in disabling RadioGroup input from virkailija-ui-components
// There are many other components also that need to be moved there. Releasing a new
// version of virkailija-ui-components at the moment is not a priority.

const Container = styled.div<{ isLast: boolean }>`
  ${({ isLast }) =>
    !isLast &&
    css`
      margin-bottom: 4px;
    `}
`;

type RadioGroupOption = { value: string; label: React.ReactNode };

export type RadioGroupProps = {
  children?: React.ReactElement<{
    checked?: boolean;
    value: string;
    onChange?: (arg: any) => void;
    disabled?: boolean;
    error?: boolean;
  }>[];
  value: string;
  onChange?: (arg: any) => void;
  disabled?: boolean;
  error?: boolean;
  options?: RadioGroupOption[];
};

export const RadioGroup = ({
  value,
  onChange,
  disabled = false,
  options,
  error = false,
  children: childrenProp,
}: RadioGroupProps) => {
  let children: React.ReactNode = null;

  if (childrenProp) {
    const validChildren = React.Children.toArray(childrenProp).filter(c =>
      React.isValidElement(c)
    );

    const childrenCount = React.Children.count(validChildren);

    children = React.Children.map(validChildren, (child, index) => {
      const checked = value !== undefined && child.props.value === value;
      const element = React.cloneElement(child, {
        checked,
        onChange,
        disabled,
        error,
      });

      const isLast = index === childrenCount - 1;

      return <Container isLast={isLast}>{element}</Container>;
    });
  } else if (isArray(options)) {
    children = options.map(({ value: optionValue, label }, index) => (
      <Container isLast={index === options.length - 1} key={optionValue}>
        <Radio
          checked={value !== undefined && value === optionValue}
          onChange={onChange}
          value={optionValue}
          error={error}
          disabled={disabled}
        >
          {label}
        </Radio>
      </Container>
    ));
  }

  return <>{children}</>;
};
