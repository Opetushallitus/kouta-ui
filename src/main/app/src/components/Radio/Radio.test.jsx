import React from 'react';

import Radio, { RadioGroup } from './index';
import { mountWithTheme } from '../../testUtils';

test('renders correctly', () => {
  expect(
    mountWithTheme(
      <RadioGroup value="b" onChange={() => {}}>
        <Radio value="a">Radio A</Radio>
        <Radio value="b">Radio B</Radio>
        <Radio value="c">Radio C</Radio>
      </RadioGroup>
    )
  ).toMatchSnapshot();
});
