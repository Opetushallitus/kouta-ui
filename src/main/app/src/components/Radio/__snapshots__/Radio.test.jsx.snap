// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  .c2 {
  position: relative;
  -webkit-flex: 0;
  -ms-flex: 0;
  flex: 0;
  line-height: 0;
  top: 0.15em;
}

.c8 {
  background-color: white;
  border-radius: 50%;
  width: 0.4em;
  height: 0.4em;
}

.c4 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.c6 {
  position: relative;
  display: inline-block;
  position: relative;
  width: 1em;
  height: 1em;
  background-color: white;
  border: 1px solid #ced4da;
  box-shadow: 0 0 0 0 transparent;
  border-radius: 50%;
  -webkit-transition: box-shadow 0.25s,border-color 0.25s,background-color 0.25s;
  transition: box-shadow 0.25s,border-color 0.25s,background-color 0.25s;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c6:hover,
.c3:focus + .c6 {
  border-color: #159ecb;
}

.c3:focus + .c6 {
  box-shadow: 0 0 0 3px rgba(21,158,203,0.3);
}

.c6 .c7 {
  visibility: hidden;
}

.c10:hover,
.c3:focus + .c10 {
  border-color: #159ecb;
}

.c3:focus + .c10 {
  box-shadow: 0 0 0 3px rgba(21,158,203,0.3);
}

.c10 .c7 {
  visibility: visible;
}

.c1 {
  cursor: pointer;
  font-family: 'Roboto',sans-serif;
  font-size: 1rem;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  line-height: 1.5;
  color: #666666;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c1:hover .c5 {
  border-color: #159ecb;
}

.c9 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin-left: 9px;
}

.c0 {
  margin-bottom: 4px;
}

<styled.div
    isLast={false}
    key=".$.0"
  >
    <div
      className="c0"
    >
      <ForwardRef
        checked={false}
        disabled={false}
        error={false}
        key=".0"
        onChange={[Function]}
        value="a"
      >
        <styled.label
          disabled={false}
          error={false}
        >
          <label
            className="c1"
            disabled={false}
          >
            <styled.div>
              <div
                className="c2"
              >
                <styled.input
                  checked={false}
                  disabled={false}
                  onChange={[Function]}
                  value="a"
                >
                  <input
                    checked={false}
                    className="c3 c4"
                    disabled={false}
                    onChange={[Function]}
                    type="radio"
                    value="a"
                  />
                </styled.input>
                <styled.div
                  checked={false}
                  error={false}
                >
                  <div
                    checked={false}
                    className="c5 c6"
                  >
                    <styled.svg>
                      <svg
                        className="c7 c8"
                      />
                    </styled.svg>
                  </div>
                </styled.div>
              </div>
            </styled.div>
            <styled.div>
              <div
                className="c9"
              >
                Radio A
              </div>
            </styled.div>
          </label>
        </styled.label>
      </ForwardRef>
    </div>
  </styled.div>,
  .c2 {
  position: relative;
  -webkit-flex: 0;
  -ms-flex: 0;
  flex: 0;
  line-height: 0;
  top: 0.15em;
}

.c8 {
  background-color: white;
  border-radius: 50%;
  width: 0.4em;
  height: 0.4em;
}

.c4 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.c10:hover,
.c3:focus + .c10 {
  border-color: #159ecb;
}

.c3:focus + .c10 {
  box-shadow: 0 0 0 3px rgba(21,158,203,0.3);
}

.c10 .c7 {
  visibility: hidden;
}

.c6 {
  position: relative;
  display: inline-block;
  position: relative;
  width: 1em;
  height: 1em;
  background-color: white;
  border: 1px solid #ced4da;
  box-shadow: 0 0 0 0 transparent;
  border-radius: 50%;
  -webkit-transition: box-shadow 0.25s,border-color 0.25s,background-color 0.25s;
  transition: box-shadow 0.25s,border-color 0.25s,background-color 0.25s;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #159ecb;
  border-color: #159ecb;
}

.c6:hover,
.c3:focus + .c6 {
  border-color: #159ecb;
}

.c3:focus + .c6 {
  box-shadow: 0 0 0 3px rgba(21,158,203,0.3);
}

.c6 .c7 {
  visibility: visible;
}

.c1 {
  cursor: pointer;
  font-family: 'Roboto',sans-serif;
  font-size: 1rem;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  line-height: 1.5;
  color: #666666;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c1:hover .c5 {
  border-color: #159ecb;
}

.c9 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin-left: 9px;
}

.c0 {
  margin-bottom: 4px;
}

<styled.div
    isLast={false}
    key=".$.1"
  >
    <div
      className="c0"
    >
      <ForwardRef
        checked={true}
        disabled={false}
        error={false}
        key=".1"
        onChange={[Function]}
        value="b"
      >
        <styled.label
          disabled={false}
          error={false}
        >
          <label
            className="c1"
            disabled={false}
          >
            <styled.div>
              <div
                className="c2"
              >
                <styled.input
                  checked={true}
                  disabled={false}
                  onChange={[Function]}
                  value="b"
                >
                  <input
                    checked={true}
                    className="c3 c4"
                    disabled={false}
                    onChange={[Function]}
                    type="radio"
                    value="b"
                  />
                </styled.input>
                <styled.div
                  checked={true}
                  error={false}
                >
                  <div
                    checked={true}
                    className="c5 c6"
                  >
                    <styled.svg>
                      <svg
                        className="c7 c8"
                      />
                    </styled.svg>
                  </div>
                </styled.div>
              </div>
            </styled.div>
            <styled.div>
              <div
                className="c9"
              >
                Radio B
              </div>
            </styled.div>
          </label>
        </styled.label>
      </ForwardRef>
    </div>
  </styled.div>,
  .c1 {
  position: relative;
  -webkit-flex: 0;
  -ms-flex: 0;
  flex: 0;
  line-height: 0;
  top: 0.15em;
}

.c7 {
  background-color: white;
  border-radius: 50%;
  width: 0.4em;
  height: 0.4em;
}

.c3 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.c5 {
  position: relative;
  display: inline-block;
  position: relative;
  width: 1em;
  height: 1em;
  background-color: white;
  border: 1px solid #ced4da;
  box-shadow: 0 0 0 0 transparent;
  border-radius: 50%;
  -webkit-transition: box-shadow 0.25s,border-color 0.25s,background-color 0.25s;
  transition: box-shadow 0.25s,border-color 0.25s,background-color 0.25s;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c5:hover,
.c2:focus + .c5 {
  border-color: #159ecb;
}

.c2:focus + .c5 {
  box-shadow: 0 0 0 3px rgba(21,158,203,0.3);
}

.c5 .c6 {
  visibility: hidden;
}

.c9:hover,
.c2:focus + .c9 {
  border-color: #159ecb;
}

.c2:focus + .c9 {
  box-shadow: 0 0 0 3px rgba(21,158,203,0.3);
}

.c9 .c6 {
  visibility: visible;
}

.c0 {
  cursor: pointer;
  font-family: 'Roboto',sans-serif;
  font-size: 1rem;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  line-height: 1.5;
  color: #666666;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c0:hover .c4 {
  border-color: #159ecb;
}

.c8 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin-left: 9px;
}

<styled.div
    isLast={true}
    key=".$.2"
  >
    <div
      className=""
    >
      <ForwardRef
        checked={false}
        disabled={false}
        error={false}
        key=".2"
        onChange={[Function]}
        value="c"
      >
        <styled.label
          disabled={false}
          error={false}
        >
          <label
            className="c0"
            disabled={false}
          >
            <styled.div>
              <div
                className="c1"
              >
                <styled.input
                  checked={false}
                  disabled={false}
                  onChange={[Function]}
                  value="c"
                >
                  <input
                    checked={false}
                    className="c2 c3"
                    disabled={false}
                    onChange={[Function]}
                    type="radio"
                    value="c"
                  />
                </styled.input>
                <styled.div
                  checked={false}
                  error={false}
                >
                  <div
                    checked={false}
                    className="c4 c5"
                  >
                    <styled.svg>
                      <svg
                        className="c6 c7"
                      />
                    </styled.svg>
                  </div>
                </styled.div>
              </div>
            </styled.div>
            <styled.div>
              <div
                className="c8"
              >
                Radio C
              </div>
            </styled.div>
          </label>
        </styled.label>
      </ForwardRef>
    </div>
  </styled.div>,
]
`;
